### PR TITLE
[bndtools] Gracefully handle missing cnf

### LINKED ISSUE #4203: [bndtools] Bndtools should mark error for projects outside of the Bnd workspace
I usually have my Eclipse workspace (containing the workspace metadata) in a different directory to my Bnd workspace. I'm happy to take advice if people feel that this is not the best approach, but it has some advantages. One drawback of this approach, however, is that when you copy a project in your Eclipse workspace, by default it copies it into the Eclipse workspace and not the Bnd workspace. 

Things don't work properly when you do this - in particular, the Bndtools builder won't generate the artifacts. But there is no direct mention of the fact that this is an issue.

In my case I did have a few errors come up like this one:
```
Error in Bnd resolution analysis.
java.lang.IllegalArgumentException: A Jar can only accept a file or directory that exists: C:\workspace\idempiere-test-workspace\org.idempiere.test.assertj\generated\org.idempiere.test.jar
	at aQute.bnd.osgi.Jar.<init>(Jar.java:138)
	at aQute.bnd.osgi.Jar.<init>(Jar.java:202)
	at bndtools.tasks.JarFileCapReqLoader.getBuilder(JarFileCapReqLoader.java:21)
	at bndtools.tasks.BndBuilderCapReqLoader.load(BndBuilderCapReqLoader.java:56)
	at bndtools.tasks.BndBuilderCapReqLoader.loadCapabilities(BndBuilderCapReqLoader.java:92)
	at bndtools.tasks.AnalyseBundleResolutionJob.run(AnalyseBundleResolutionJob.java:61)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
```
However, this error does not directly mention the root cause of the problem. Because my eyes are normally drawn to the end of the path (which also had an issue), it was a while before I noticed that it was in the `C:\workspace\idempiere-test-workspace` directory when it should have been in the `C:\workspace\idempiere-test` directory (where the Bnd workspace is). 

It should be a fairly simple matter for Bndtools to place some kind of error marker on a project when it finds that it is not in the Bnd workspace - that would have saved me a good hour or so trying to track down the cause of this issue.

### LINKED ISSUE #4401: [bndtools] Bndtools doesn't handle cnf project created after starting
* Start with a normal Bndtools workspace in Eclipse.
* Delete the cnf project (without deleting the contents on disk). The Bndtools Builder will get into an annoying infinite loop with build errors, with a Build Error dialog popping up stating that the cnf project is missing. (While this could be handled more gracefully, that is a separate issue to the one I'm describing in this issue.)
* Re-import the cnf project. Annoying infinite loop and Build Error dialog stops.
* However, now every project's build is failing with an error marker (among others) stating "Bndtools: Error in setup, likely the cnf folder is not ../cnf relative to the project folder '&lt;path to project&gt;'. The workspace is in '&lt;path to workspace&gt;'.
* This situation seems to persists until you restart Eclipse, then the build starts normally.

Admittedly, this is not a common operation in ordinary development - you normally create your cnf project at the start and it remains for the life of the workspace. However, I believe it may be the root cause of #4253, or at the very least closely related - the integration test, in the `@BeforeAll` phase, clears the workspace of all projects (including cnf) and then re-imports them. My experience through the above demonstrates that there is a race condition in Bndtools such that it doesn't always react perfectly to the disappearance and reappearance of the cnf project, and it ends up remembering that the cnf project is missing. The code in the Bndtools builder that generates that error generates it when `Central` returns `null` when trying to get the Bnd `Project` object:

https://github.com/bndtools/bnd/blob/a45a1820b17b0604474f85bc7aeaaf9b40880ae6/bndtools.builder/src/org/bndtools/builder/validate/ProjectPathsValidator.java#L77-L89

This is the same code path that triggers when the integration test fails.

I believe that this may also have been the root cause of #4202. I think at the time that that happened, I was messing around with my cnf and may have deleted/reimported it. It looks like a similar error.

### PR DESCRIPTION
**Pull request by kriegfrj**
Without this fix, Bndtools generates an NPE during building if the cnf project is missing and continues to fail building even after the cnf project is added. This fixes #4401.

As a side effect, also introduces an error marker if the Bnd project is not inside the Bnd workspace. This fixes #4203.

This seems to work in my testing and not to break anything else, but it's hard to be sure without a decent set of regression tests. It would be good for us to smoke-test it on dev for a while before release.

### CODE REVIEW

File: bndtools.builder/src/org/bndtools/builder/BndtoolsBuilder.java
Comment (bjhargrave) [inline bndtools.builder/src/org/bndtools/builder/BndtoolsBuilder.java:122]: You now call `markers.deleteMarkers("*")` twice if you get an exception on `Central.getProject(myProject)`.

File: bndtools.builder/src/org/bndtools/builder/MarkerSupport.java
Comment (bjhargrave) [inline bndtools.builder/src/org/bndtools/builder/MarkerSupport.java:101]: Should this not be `cnf.isOpen()`?

Comment (kriegfrj) [reply bndtools.builder/src/org/bndtools/builder/MarkerSupport.java:101]: I don't know enough about Eclipse's marker handling to know the answer to this question. Checking for existence was sufficient to stop the NPE from happening in the enclosing code if the cnf project was missing. But checking for openness is probably not a bad idea.

File: bndtools.builder/src/org/bndtools/builder/BndtoolsBuilder.java
Comment (kriegfrj) [reply bndtools.builder/src/org/bndtools/builder/BndtoolsBuilder.java:137]: Fixed.

**Comment by kriegfrj** [comment]
Ok, latest push fixes the issues raised by @bjhargrave and adds a number of other enhancements:

1. If there is an error at this level for a particular project, then not only are all the Bnd markers deleted, but all the Java markers as well. The reason for this: If your cnf is missing/misconfigured in some way, there will inevitably be build path problems that will lead to dozens or hundreds of compile errors in your project. Not only are these errors not particularly helpful (as they don't point directly to the root cause), there are so many of them that the root cause can get lost in the noise. So now we remove them to make the root cause more prominent.
2. The builder now performs some more precise diagnostic tests once the missing `Project` is detected, in order to give a more detailed and specific report as to the possible reason for the failure.

@bjhargrave, I'm also hopeful that this will fix #4253. If you can keep an eye on the integration test after this PR is merged and if that failure stops popping up maybe we can re-instate the integration test. I'll also look at adding regression tests for the fixes in this PR if that's the case.

File: bndtools.builder/src/org/bndtools/builder/BndtoolsBuilder.java
Comment (bjhargrave) [inline bndtools.builder/src/org/bndtools/builder/BndtoolsBuilder.java:130]: It does seem a bit presumptuous of us (Bndtools) to remove the error markers of another builder, no? I am not sure we should do this.

File: bndtools.builder/src/org/bndtools/builder/BndtoolsBuilder.java
Comment (bjhargrave) [inline bndtools.builder/src/org/bndtools/builder/BndtoolsBuilder.java:142]: It does seem a bit presumptuous of us (Bndtools) to remove the error markers of another builder, no? I am not sure we should do this.

File: bndtools.builder/src/org/bndtools/builder/BndtoolsBuilder.java
Comment (kriegfrj) [reply bndtools.builder/src/org/bndtools/builder/BndtoolsBuilder.java:130]: I suppose i could try and see if we can coerce the JavaBuilder to do it itself under these conditions, eg by failing the BndClasspathContainer somehow.

Comment (bjhargrave) [reply bndtools.builder/src/org/bndtools/builder/BndtoolsBuilder.java:130]: I don't think that is a good idea or possible either. The Java builder's errors are real. They are compilation failures. And they are the responsibility of the Java builder to manage.

Comment (kriegfrj) [reply bndtools.builder/src/org/bndtools/builder/BndtoolsBuilder.java:130]: I disagree, but rather than hold up this PR over this issue (which is not directly related to the original problem)  I have removed the offending lines for now so that we can get it merged. We can discuss this further at another time.

**Comment by pkriens** [comment]
Just a comment in this area ... why not use the Bndtools Explorer to report about this case? The Bndtools Explorer is really about the Workspace model and it can easily provide a lot of information that cnf is missing.

**Comment by kriegfrj** [comment]
@bjhargrave, are you still waiting for me to do something before meeting this?

**Comment by bjhargrave** [comment]
The build is bad

**Comment by kriegfrj** [comment]
The failure is in bndlib tests. I can't see how that was caused by this PR?

**Comment by bjhargrave** [comment]
yes, but it is in another test you authored which is now regularly failing CI builds like this one.

https://github.com/bndtools/bnd/pull/4404/checks?check_run_id=1389155797#step:5:449

Can you look into this? I would like to stabilize the CI builds.

**Review by bjhargrave — APPROVED** [review]

**Pull request merged**
